components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ImmutablePair.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.config.provider;
+
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * An {@link ImmutablePair} consists of two elements within. The elements once set
+ * in the ImmutablePair cannot be modified. The class itself is final, so that it
+ * cannot be subclassed. This is general norm for creating Immutable classes.
+ * Please note that the {@link ImmutablePair} cannot be modified once set, but the
+ * objects within them can be, so in general it means that if there are mutable objects
+ * within the pair then the pair itself is effectively mutable.
+ * <p>
+ * <pre>
+ *     ImmutablePair<Tuple, TupleInputStreamIdentifier> tupleStreamPair =
+ *      new ImmutablePair<Tuple, TupleInputStreamIdentifier> (tuple, identifier);
+ *   ...
+ *   ...
+ *   Tuple t = tupleStreamPair.getFirst();
+ *   TupleInputStreamIdentifier identifier = tupleStreamPair.getSecond();
+ *   ...
+ * </pre>
+ *
+ * @param <A> type A
+ * @param <B> type B
+ * @since 2.0.6
+ */
+final class ImmutablePair<A, B> {
```

Comment:
Whats the purpose of the class and who uses this?

Reply:
This is used in [1], where we return the unique attribute field name and its value.

Ex:- 
```
TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_NAME="abc"
TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_PORT="8000"
```

In here, the unique identifier's env variable is `TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_NAME`. If you pass `TESTCONFIGURATION_TRANSPORTS_TRANSPORT_0_PORT` in the method [1], you will get ImmutablePair.of("NAME", "abc")

We do need the name and the value when we need to identify whether a configuration needs to be overridden or added (ex: add a new transport via environment variables)

[1] https://github.com/wso2/carbon-config/pull/54/files#diff-a47faff6be5f4d2f88ebc8088d7a1c29R380